lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -612,39 +701,304 @@ void ScienceSensorsSystemPrivate::GenerateOctrees()
+    this->timeSpaceCoords[this->timeIdx]->points.end(),
+    zSlice1.points.begin(),
+    zSlice1.points.end(),
+    std::inserter(ptsExceptZSlice1, ptsExceptZSlice1.begin()),
+    std::bind(&ScienceSensorsSystemPrivate::comparePclPoints, this,
+      std::placeholders::_1, std::placeholders::_2));
+
+  // Step 3: Look for 2nd NN everywhere except z slice of 1st NN.
+  // In this 2nd z-slice, search for 4 NNs
+
+  // Create point cloud from set difference result
+  pcl::PointCloud<pcl::PointXYZ> cloudExceptZSlice1;
+  cloudExceptZSlice1.points.insert(cloudExceptZSlice1.points.begin(),
+    ptsExceptZSlice1.begin(), ptsExceptZSlice1.end());
+
+  // Create octree
+  auto octreeExceptZSlice1 = pcl::octree::OctreePointCloudSearch<pcl::PointXYZ>(
+    this->spatialRes);
+  octreeExceptZSlice1.setInputCloud(cloudExceptZSlice1.makeShared());
+  octreeExceptZSlice1.addPointsFromInputCloud();
+
+  // Search results
+  std::vector<int> inds2;
+  std::vector<float> sqrDists2;
+  int nnIdx2 = -1;
+  float minDist2 = -1.0f;
+  float nnZ2 = -1.0f;
+
+  // Look for 2nd NN
+  if (octreeExceptZSlice1.getLeafCount() > 0)
+  {
+    if (octreeExceptZSlice1.nearestKSearch(_pt, 1, inds2, sqrDists2)
+      <= 0)
+    {
+      ignwarn << "Cannot find 2nd NN for trilinear interpolation" << std::endl;
+      return;
+    }
+    else
+    {
+      // Take closest point as 2nd NN
+      nnIdx2 = inds2[0];
+      minDist2 = sqrDists2[0];
+      // Get z of neighbor
+      nnZ2 = cloudExceptZSlice1.at(nnIdx2).z;
+
+      igndbg << "Slice 2 nnIdx " << nnIdx2 << ", dist " << sqrt(minDist2)
+        << std::endl;
+    }
+  }
+
+  // Step 4: Look for 4 NNs in the z slice of 2nd NN
+
+  // Get raw Eigen matrix
+  // Matrix size 3 x n
+  Eigen::MatrixXf matrixExceptZSlice1 =
+    cloudExceptZSlice1.getMatrixXfMap(3, 4, 0).block(
+      0, 0, 3, cloudExceptZSlice1.size());
+
+  // Search in z slice of 1st NN for 4 nearest neighbors in this slice
+  this->SearchInDepthSlice(_pt, nnZ2, matrixExceptZSlice1, zSlice2, zSliceInds2,
+    interpolatorInds2, interpolatorSqrDists2, _interpolators2);
+  if (interpolatorInds2.size() < 4 || interpolatorSqrDists2.size() < 4)
+  {
+    ignwarn << "Could not find enough neighbors in 2nd slice z = " << nnZ
+      << " for trilinear interpolation." << std::endl;
+  }
+}
+
+/////////////////////////////////////////////////
+void ScienceSensorsSystemPrivate::SearchInDepthSlice(
+  pcl::PointXYZ &_searchPt,
+  float _depth,
+  Eigen::MatrixXf &_points,
+  pcl::PointCloud<pcl::PointXYZ> &_zSlice,
+  std::vector<int> &_zSliceInds,
+  std::vector<int> &_interpolatorInds,
+  std::vector<float> &_interpolatorSqrDists,
+  std::vector<pcl::PointXYZ> &_interpolators,
+  int _k)
+{
+  // Find indices of points with z matching the given z. These points are on
+  // the same z slice. Logical indexing, produces Boolean array.
+  // Take only the block of the matrix containing the z values of the points.
+  Eigen::Matrix<bool, 1, Eigen::Dynamic> depthInds =
```

Comment:
Floating point equality comparisions are very fragile. We shouldn't rely on them here. 

There's PCL functionality that I think will do this much more cleanly: https://pcl.readthedocs.io/projects/tutorials/en/latest/passthrough.html#passthrough and hopefully efficiently.

Reply:
7c85554 refactors to use a PassThrough filter.

Here I did a `==` for speedy implementation, because the data we have is very clean with integers. The quick alternative is to instead do `abs(a-b) < 1e-6`, which is pretty standard.

PassThrough filter still requires specifying a min and max range, which means we still need to do the same thing - pass in thresholds like `1e-6`.

In its defense, the code here actually reimplements PassThrough filter at a lower level, with the added flexibility that we can give it any if-condition we want, not limited to a min/max range in a specified field. In the new code, using the filter allows for simpler code, but loses that low-level control.